keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1076,361 @@ the remaining requests in that queue start getting faster service.  In
+equal to `<width> x <processing latency>`.  The rationale behind it is that
+the request is now consuming `<width>` concurrency units for the duration
+of its processing.
+
+While in theory the `width` can be an arbitrary non-integer number, for
+practical reasons, we will assume it actually is an integer.  Given that
+our estimations here are very rough anyway that seems a reasonable
+simplification that makes dispatching the budget a bit simpler.
+
+#### Determining the width
+
+While one can imagine arbitrarily sophisticated algorithms for it (including
+exposing defining the width of requests via FlowSchema API), we want to start
+with something relatively simple to first get operational experience with it
+before investing into sophisticated algorithms or exposing a knob to users.
+
+In order to determine the function that will be approximating the `width` of
+a request, we should first estimate how expensive a particular request is.
+And we need to think about both dimensions that we're trying to protect from
+overloading (CPU and RAM) and how many concurrency units a request can actually
+consume.
+
+Let's start with CPU.  The total cost of processing a LIST request should be
+proportional to the number of processed objects.  However, given that in
+practice processing a single request isn't parallelized (and the fact that
+we generally scale the number of total concurrency units linearly with amount
+of available resources), a single request should consume no more than A
+concurrency units.  Fortunately that all compiles together because the
+`processing latency` of the LIST request is actually proportional to the
+number of processed objects, so the cost of the request (defined above as
+`<width> x <processing latency>` really is proportaional to the number of
+processed objects as expected.
+
+For RAM the situation is actually different.  In order to process a LIST
+request we actually store all objects that we process in memory. Given that
+memory is uncompressable resource, we effectively need to reserve all that
+memory for the whole time of processing that request.  That suggests that
+the `width` for the request from the RAM perspective should be proportional
+to the number of processed items.
+
+So what we get is that:
+```
+  width_cpu(N) = min(A, B * N)
+  width_ram(N) = C * N
+```
+where N is the number of items a given LIST request is processing.
+
+The question is how to combine them to a single number.  While the main goal
+is to stay on the safe side and protect from the overload, we also want to
+maxiumize the utilization of the available concurrency units.
+Fortunately, when we normalize CPU and RAM to percentage of available capacity,
+it appears that almost all requests are much more cpu-intensive.  Assuming
+4GB:1CPU ratio and 10kB average object and the fact that processing larger
+number of objects can utilize exactly 1 core, that means that we need to
+process 400.000 objects to make the memory cost higher.
```

Comment:
This does not feel like the right ratio.  Does it really match up with experience?
Remember that a non-LIST request does not utilize a CPU for the whole time it is running.
Every sort of request takes only a factional bite of CPU.
_Maybe_ the faction varies from sort to sort.

Reply:
> Remember that a non-LIST request does not utilize a CPU for the whole time it is running.

True. But as i described above, 1 seat doesn't correspond  to 1 cpu. We generally have a relation of "N concurrency units per core". So with N in-flight limit in practice we can saturate 1 core and similarly we can saturate with 1 large enough LIST request. It's hard to say if in practice we observe 1:400k ratio, but we definitely observe the orders of magnitude difference.

Reply:
Now we have different uses of the symbol `N`.  Let me try to restate with distinct variables.
For normal requests, the only kind we have now, we use a concurrency limit of 600 by default.  That is the limit used regardless of the number of local cores.  BTW, we probably do not even really mean cores here, as Linux and Golang deal more prominently with "CPUs", of which we get two per core on a typical Intel machine.  Let's take 16 as a typical number of CPUs and work with that.  So before introducing LISTs, we are supposing that the amortized CPU cost of one seat is about 0.027 CPUs.  You said the memory cost of that same seat is about 10 KB.  To fully occupy one CPU with non-LIST requests takes 37.5 seats, which corresponds to 375 KB of memory.  It would take a memory inflation factor of about 11,000 to reach 4 GB of memory.

Reply:
> For normal requests, the only kind we have now, we use a concurrency limit of 600 by default. That is the limit used regardless of the number of local cores.

This is the default. I can't imagine any reasonable operator not adjusting it. There is no way you can reasonable handle 600 inflight limits if your master e.g. has 2 cores. 

>  Let's take 16 as a typical number of CPUs and work with that. So before introducing LISTs, we are supposing that the amortized CPU cost of one seat is about 0.027 CPUs. You said the memory cost of that same seat is about 10 KB.

Those are your numbers. I wasn't involved in setting defaults and those aren't the numbers that we use (I meant we don't have 600 limits for 16-core master). So this discussion is moot.